packages/gatsby-plugin-sharp/src/index.js

```diff
@@ -93,62 +132,157 @@ function queueImageResizing({ file, args = {}, reporter }) {
-  const job = {
-    args: options,
-    inputPath: file.absolutePath,
-    contentDigest: file.internal.contentDigest,
-    outputDir,
-    outputPath: outputFilePath,
+  return {
+    src: prefixedSrc,
+    outputDir: outputDir,
+    relativePath: outputFilePath,
+    width,
+    height,
+    aspectRatio,
+    options: removeDefaultValues(args, getPluginOptions()),
   }
+}
 
-  const outputFile = path.join(job.outputDir, job.outputPath)
-  let finishedPromise = Promise.resolve()
+function createJob(job, { reporter }) {
+  const progressBar = createOrGetProgressBar(reporter)
 
-  if (cachedOutputFiles.has(outputFile)) {
-    finishedPromise = cachedOutputFiles.get(outputFile)
+  if (pendingImagesCounter === 0) {
+    progressBar.start()
   }
 
-  // Check if the output file already exists or already is being created.
-  // TODO: Remove this when jobs api is stable, it will have a better check
-  if (!fs.existsSync(outputFile) && !cachedOutputFiles.has(outputFile)) {
-    // schedule job immediately - this will be changed when image processing on demand is implemented
-    finishedPromise = scheduleJob(
-      job,
-      boundActionCreators,
-      pluginOptions,
-      reporter
-    ).then(res => {
-      cachedOutputFiles.delete(outputFile)
+  const transformsCount = job.args.operations.length
+  pendingImagesCounter += transformsCount
+  progressBar.total = pendingImagesCounter
+
+  // Jobs can be duplicates and usually are long running tasks.
+  // Because of that we shouldn't use async/await and instead opt to use
+  // .then() /.catch() handlers, because this allows V8 to release
+  // duplicate jobs from memory quickly (as job is not referenced
+  // in resolve / reject handlers). If we would use async/await
+  // entire closure would keep duplicate job in memory until
+  // initial job finish.
+  let promise = null
+  if (boundActionCreators.createJobV2) {
+    promise = boundActionCreators.createJobV2(job)
+  } else {
+    promise = scheduleJob(job, boundActionCreators)
+  }
 
-      return res
+  promise
+    .catch(err => {
+      reporter.panic(err)
+    })
+    .then(() => {
+      progressBar.tick(transformsCount)
     })
 
-    cachedOutputFiles.set(outputFile, finishedPromise)
-  }
+  return promise
+}
+
+function queueImageResizing({ file, args = {}, reporter }) {
+  const fullOptions = healOptions(getPluginOptions(), args, file.extension)
+  const {
+    src,
+    width,
+    height,
+    aspectRatio,
+    relativePath,
+    outputDir,
+    options,
+  } = prepareQueue({ file, args: createTransformObject(fullOptions) })
+
+  // Create job and add it to the queue, the queue will be processed inside gatsby-node.js
+  const finishedPromise = createJob(
+    {
+      name: IMAGE_PROCESSING_JOB_NAME,
+      inputPaths: [file.absolutePath],
+      outputDir,
+      args: {
+        operations: [
+          {
+            outputPath: relativePath,
+            args: removeDefaultValues(options, getPluginOptions()),
```

Comment:
do we need to call `removeDefaultValues()` here? it does seem to be called on `options` in `prepareQueue` already